go/libraries/doltcore/table/typed/parquet/writer.go

```diff
@@ -0,0 +1,137 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parquet
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/xitongsys/parquet-go-source/local"
+	"github.com/xitongsys/parquet-go/source"
+	"github.com/xitongsys/parquet-go/writer"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/row"
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema/typeinfo"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/sqlutil"
+)
+
+var WriteBufSize = 256 * 1024
+
+type ParquetWriter struct {
+	filewriter source.ParquetFile
+	pwriter    *writer.CSVWriter
+	sch        schema.Schema
+}
+
+func NewParquetWriter(outSch schema.Schema, destName string) (*ParquetWriter, error) {
+	allCols := outSch.GetAllCols()
+	columns := allCols.GetColumns()
+
+	ColTypeMap := getTypeMap()
+	var csvSchema []string
+	var repetitionType string
+	for _, col := range columns {
+		colName := col.Name
+		colType := col.TypeInfo.GetTypeIdentifier()
+		if col.IsNullable() {
+			repetitionType = `, repetitiontype=OPTIONAL`
+		} else {
+			repetitionType = ``
+		}
+		csvSchema = append(csvSchema, fmt.Sprintf("name=%s, %s%s", colName, ColTypeMap[colType], repetitionType))
+	}
+
+	fw, err := local.NewLocalFileWriter(destName)
+	if err != nil {
+		return nil, err
+	}
+
+	// np configures the degree of concurrency for our Reader and Writers
+	// TODO: not sure what default value to set 'np' to
+	pw, err := writer.NewCSVWriter(csvSchema, fw, 4)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ParquetWriter{filewriter: fw, pwriter: pw, sch: outSch}, nil
+}
+
+func (pwr *ParquetWriter) GetSchema() schema.Schema {
+	return pwr.sch
+}
+
+// WriteRow will write a row to a table
+func (pwr *ParquetWriter) WriteRow(ctx context.Context, r row.Row) error {
+	allCols := pwr.sch.GetAllCols()
+	colValStrs := make([]*string, allCols.Size())
+
+	sqlRow, err := sqlutil.DoltRowToSqlRow(r, pwr.GetSchema())
+	if err != nil {
+		return err
+	}
+
+	for i, val := range sqlRow {
+		if val == nil {
+			colValStrs[i] = nil
+		} else {
+			v := sqlutil.SqlColToStr(ctx, val)
+			colValStrs[i] = &v
+		}
+	}
+
+	err = pwr.pwriter.WriteString(colValStrs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close should flush all writes, release resources being held
+func (pwr *ParquetWriter) Close(ctx context.Context) error {
+	// WriteStop writes footer, stops writing and flushes
+	err := pwr.pwriter.WriteStop()
+	if err != nil {
+		return err
+	}
+	pwr.filewriter.Close()
+	return nil
+}
+
+func getTypeMap() map[typeinfo.Identifier]string {
```

Comment:
See if these can be extended to the logical types: https://github.com/xitongsys/parquet-go#logical-type

Reply:
You don't have to make this a method. You can define this map as a var in the top of the file. You are essentially creating this map in memory every time WriteRow is called.